File: NeuralNetwork/GradientDescentInitial.cs
// using System;
// using System.Collections.Generic;
// using System.IO;
// using System.Linq;

// var lines = File.ReadAllLines("data.csv");
// List<float> Xs = new List<float>();
// List<float> Ys = new List<float>();
// float learningRate = 0.01f;
// int iterations = 10000;

// Func<float, float, float, float> linearModel = (theta1, X, theta0) => theta1*X + theta0;

// float theta0 = 0;
// float theta1 = 0;


// foreach (var line in lines)
// {
//     var parts = line.Split(',');
//     Xs.Add(float.Parse(parts[0]));
//     Ys.Add(float.Parse(parts[1]));
// }

// for (int i = 0; i < iterations; i++)
// {
//     var loss = CalculateLoss();
//     var gradientTheta0 = GetGradientTheta0();
//     var gradientTheta1 = GetGradientTheta1();
//     Console.WriteLine($"Iteration: {i + 1}, linear model: {theta1}x + {theta0}, Loss: {loss}, GradientTheta0: {gradientTheta0}, GradientTheta1: {gradientTheta1}");
//     theta0 -= learningRate * gradientTheta0;
//     theta1 -= learningRate * gradientTheta1;
// }


// Console.WriteLine("Training done, pass input X value to get predictions: ");
// string input = Console.ReadLine();

// var result = linearModel(theta1, float.Parse(input), theta0);
// //result = result * stdDevY + meanY; //denormalize the output
// Console.WriteLine($"Prediction for input {input}: {result}");

// float GetGradientTheta0()
// {
//     float gradientValueTheta0 = 0;
//     for (int i = 0; i < Xs.Count; i++)
//     {
//         gradientValueTheta0 += (linearModel(theta1, Xs[i], theta0) - Ys[i]);
//     }
//     gradientValueTheta0 /= Xs.Count;

//     return gradientValueTheta0;
// }

// float GetGradientTheta1()
// {
//     float gradientValueTheta1 = 0;
//     for (int i = 0; i < Xs.Count; i++)
//     {
//         gradientValueTheta1 += (linearModel(theta1, Xs[i], theta0) - Ys[i]) * Xs[i];
//     }
//     gradientValueTheta1 /= Xs.Count;

//     return gradientValueTheta1;
// }

// float CalculateLoss()
// {
//     float totalLoss = 0;
//     for (int i = 0; i < Xs.Count; i++)
//     {
//         float MSE = MathF.Pow(linearModel(theta1, Xs[i], theta0) - Ys[i], 2);
//         totalLoss += MSE;
//     }
//     totalLoss /= Xs.Count;

//     return totalLoss;
// }

File: NeuralNetwork/Python/GradientDescentInitial.py
# Author: Justin Farley
# 
# This file was my initial understanding of basic Gradient Descent 
# before tackling neural networks themselves.
# Since this gradient descent weight update logic is used in neural networks
# this was a good starting point. It was also very cool to see it "learn" the equation
# (it was even cooler seeing it learn the values themselves <see cref="Neural_Network_and_AI.NeuralNetwork"/>)

import csv

# Read data
Xs = []
Ys = []

learningRate = 0.0000001
iterations = 2000000

theta0 = 0.0
theta1 = 0.0


def linearModel(theta1, X, theta0):
    return theta1 * X + theta0


# Load CSV
with open("data.csv", "r") as file:
    reader = csv.reader(file)
    for row in reader:
        Xs.append(float(row[0]))
        Ys.append(float(row[1]))


def GetGradientTheta0():
    gradientValueTheta0 = 0.0
    for i in range(len(Xs)):
        gradientValueTheta0 += (
            linearModel(theta1, Xs[i], theta0) - Ys[i]
        )

    gradientValueTheta0 /= len(Xs)
    return gradientValueTheta0


def GetGradientTheta1():
    gradientValueTheta1 = 0.0
    for i in range(len(Xs)):
        gradientValueTheta1 += (
            (linearModel(theta1, Xs[i], theta0) - Ys[i]) * Xs[i]
        )

    gradientValueTheta1 /= len(Xs)
    return gradientValueTheta1


def CalculateLoss():
    totalLoss = 0.0
    for i in range(len(Xs)):
        mse = (linearModel(theta1, Xs[i], theta0) - Ys[i]) ** 2
        totalLoss += mse

    totalLoss /= len(Xs)
    return totalLoss


# Training loop
for i in range(iterations):

    loss = CalculateLoss()
    gradientTheta0 = GetGradientTheta0()
    gradientTheta1 = GetGradientTheta1()

    print(
        f"Iteration: {i + 1}, "
        f"linear model: {theta1}x + {theta0}, "
        f"Loss: {loss}, "
        f"GradientTheta0: {gradientTheta0}, "
        f"GradientTheta1: {gradientTheta1}"
    )

    theta0 -= learningRate * gradientTheta0
    theta1 -= learningRate * gradientTheta1


print("Training done, pass input X value to get predictions:")
input_value = float(input())

result = linearModel(theta1, input_value, theta0)

print(f"Prediction for input {input_value}: {result}")
